worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala

```diff
@@ -1246,6 +1199,51 @@ class PushDataHandler(val workerSource: WorkerSource) extends BaseMessageHandler
     }
   }
 
+  private def writeLocalData(
+      fileWriters: Seq[FileWriter],
+      body: ByteBuf,
+      shuffleKey: String,
+      batchOffsets: Option[Array[Int]],
+      writePromise: Option[Promise[Unit]]): Unit = {
+    def writeData(fileWriter: FileWriter, body: ByteBuf, shuffleKey: String): Unit = {
+      try {
+        fileWriter.write(body)
+      } catch {
+        case e: AlreadyClosedException =>
+          fileWriter.decrementPendingWrites()
+          val (mapId, attemptId) = getMapAttempt(body)
+          val endedAttempt =
+            if (shuffleMapperAttempts.containsKey(shuffleKey)) {
+              shuffleMapperAttempts.get(shuffleKey).get(mapId)
+            } else -1
+          // TODO just info log for ended attempt
+          logWarning(s"Append data failed for task(shuffle $shuffleKey, map $mapId, attempt" +
+            s" $attemptId), caused by AlreadyClosedException, endedAttempt $endedAttempt, error message: ${e.getMessage}")
+          writePromise.map(_.failure(e)).orElse(throw e)
+        case e: Exception =>
+          logError("Exception encountered when write.", e)
+          writePromise.map(_.failure(e)).orElse(throw e)
+      }
+    }
+    batchOffsets match {
+      case Some(batchOffsets) =>
+        batchOffsets.zip(fileWriters).foreach { case (offset, writer) =>
+          val length =
+            if (offset == batchOffsets.last) {
+              body.readableBytes() - offset
+            } else {
+              batchOffsets(batchOffsets.indexOf(offset) + 1) - offset
+            }
+          val batchBody = body.slice(body.readerIndex() + offset, length)
+          writeData(writer, batchBody, shuffleKey)
+        }
+        writePromise.foreach(_.success())
```

Comment:
Seems we should only call `success` when `writePromise` is not failed in `writeLocalData`

Reply:
if `writePromise` has already failed the success call won't have any effect. It will already be in failed state anyways

Reply:
in any case, i have added a check to see if the promise has not been completed yet, and only then calling success on it

Reply:
```
Completes the promise with either an exception or a value.
Params:
result – Either the value or the exception to complete the promise with.
If the promise has already been fulfilled, failed or has timed out, calling this method will throw an IllegalStateException.
```
According to the comments, it will throw `IllegalStateException` if the promise has already been fulfilled or failed, so I think it's necessary :)

Reply:
Sounds good, I updated it anyways 🙂

Reply:
It would be better to model it as `CompletableFuture` instead - it is more closer to what we need here.